cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/StructuredTable.java

```
@@ -333,4 +334,23 @@ default long count(Collection<Range> keyRanges,
       Collection<Field<?>> filterIndexes) throws InvalidFieldException, IOException {
     throw new UnsupportedOperationException("No supported implementation.");
   }
+
+  /**
+   * Get the number of records from the table matching the key range of specified index fields, with
+   * additional query options.
+   *
+   * @param keyRanges     key ranges of the rows to count
+   * @param filterIndexes the indexes to filter upon, OR logic will be applied
+   * @param options       optional {@link QueryOption}s to modify the query behavior (e.g., stale
+   *                      reads)
+   * @return the number of records matching the criteria
+   * @throws InvalidFieldException if a field is not part of the table schema, is not an indexed
+   *                               column, or the type does not match the schema
+   * @throws IOException           if there is an error reading from the table
+   */
+  default long count(Collection<Range> keyRanges,
+      Collection<Field<?>> filterIndexes, QueryOption... options)
+      throws InvalidFieldException, IOException {
+    return count(keyRanges, filterIndexes);
```

Why not `throw new UnsupportedOperationException("No supported implementation.");` instead of calling `count(keyRanges, filterIndexes)`?

---

The method needs override only in case of Spanner implementation. In other cases (query options are not very useful right now) - we use the existing count method.

---

But that should be overriden by other implementations to explicitly define that.

---

Updated. The initial plan was this method should be overridden only where necessary hence the default implementation was added to the interface class.

---

Thanks, we should not add implementation dependent detail in the default implementation.